soc/xtensa/intel_adsp/tools/cavstool.py

```diff
@@ -0,0 +1,300 @@
+    # are configured?) the DSP hardware will hang fairly reliably.
+    log.info("Wait for ROM startup")
+    time.sleep(0.1)
+    while (dsp.SRAM_FW_STATUS >> 24) != 5: pass
+
+    # Send the DSP an IPC message to tell the device how to boot.
+    # Note: with cAVS 1.8+ the ROM receives the stream argument as an
+    # index within the array of output streams (and we always use the
+    # first one by construction).  But with 1.5 it's the HDA index,
+    # and depends on the number of input streams on the device.
+    stream_idx = hda_ostream_id if cavs15 else 0
+    ipcval = (  (1 << 31)            # BUSY bit
+                | (0x01 << 24)       # type = PURGE_FW
+                | (1 << 14)          # purge_fw = 1
+                | (stream_idx << 9)) # dma_id
+    log.info(f"Sending IPC command, HIPCR = 0x{ipcval:x}")
+    dsp.HIPCIDR = ipcval
+
+    log.info(f"Starting DMA, FW_STATUS = 0x{dsp.SRAM_FW_STATUS:x}")
+    sd.CTL |= 2 # START flag
+
+    log.info(f"Waiting for firmware handoff, FW_STATUS = 0x{dsp.SRAM_FW_STATUS:x}")
+    for _ in range(200):
+        alive = dsp.SRAM_FW_STATUS & ((1 << 28) - 1) == 5 # "FW_ENTERED"
+        if alive: break
+        time.sleep(0.01)
+    if not alive:
+        log.warning(f"Load failed?  FW_STATUS = 0x{dsp.SRAM_FW_STATUS:x}")
+
+    # Turn DMA off and reset the stream.  Clearing START first is a noop
+    # per the spec, but required for 1.5, and makes the load on 1.8
+    # unstable.  Go figure.
+    if cavs15:
+        sd.CTL &= ~2 # clear START
+    sd.CTL |= 1
+    log.info(f"cAVS firmware load complete")
+
+# This SHOULD be just "mem[start:start+length]", but slicing an mmap
+# array seems to be unreliable on one of my machines (python 3.6.9 on
+# Ubuntu 18.04).  Read out bytes individually.
+def win_read(start, length):
+    return b''.join(bar4_mmap[x + WINSTREAM_OFFSET].to_bytes(1, 'little')
+                    for x in range(start, start + length))
+
+def win_hdr():
+    return struct.unpack("<IIII", win_read(0, 16))
+
+# Python implementation of the same algorithm in sys_winstream_read(),
+# see there for details.
+def winstream_read(last_seq):
+    while True:
+        (wlen, start, end, seq) = win_hdr()
+        if seq == last_seq or start == end:
+            return (seq, "")
+        behind = seq - last_seq
+        if behind > ((end - start) % wlen):
+            return (seq, "")
+        copy = (end - behind) % wlen
+        suffix = min(behind, wlen - copy)
+        result = win_read(16 + copy, suffix)
+        if suffix < behind:
+            result += win_read(16, behind - suffix)
+        (wlen, start1, end, seq1) = win_hdr()
+        if start1 == start and seq1 == seq:
+            return (seq, result.decode("utf-8"))
+
+async def main():
+    global hda, sd, dsp, hda_ostream_id, cavs15
+    (hda, sd, dsp, hda_ostream_id, cavs15) = map_regs()
+    log.info(f"Detected cAVS {'1.5' if cavs15 else '1.8+'} hardware")
+
+    load_firmware(sys.argv[1])
```

Comment:
Allow for logging only perhaps, though perilous....

```suggestion
    if len(sys.argv) > 1:
        load_firmware(sys.argv[1])
```